client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnection.java

```diff
@@ -507,7 +507,17 @@ public boolean isValid(int timeout)
         if (timeout < 0) {
             throw new SQLException("Timeout is negative");
         }
-        return !isClosed();
+
+        if (isClosed()) {
+            return false;
+        }
+
+        try (TrinoStatement statement = doCreateStatement()) {
+            return statement.internalExecute("SELECT 1");
```

Comment:
Can we verify the connection without polluting cluster query history?

Reply:
Not sure. Is there an existing endpoint we can use?

Reply:
I think @dain or @electrum may know.

Reply:
I've looked through endpoints annotated with `@ResourceSecurity(AUTHENTICATED_USER)`.  `/v1/query/{queryId}` could work (we could pass a dummy query id and consider 410 Gone a success), but feels hacky.  The alternative would be adding a new endpoint, though that would mean this functionality would only work with newer versions of Trino.

One thing to consider with the `SELECT 1` approach is that it could potentially introduce latency (one extreme case would be that the query gets queued because of high load or resource group configuration).

Reply:
Thanks for your suggestion. Regarding the latency, the previous code verified whether it's closed or not, so users can switch to `Connection.isClosed` method in such cases. It requires their code change and may not work in BI tools though.

Reply:
@dain @electrum Do you have any suggestion to verify JDBC connection?

Reply:
`isValid` is not something that gets called so frequently as to cause issues with queing. Also I would argue that from a user's perspective the result of `isValid` is to know whether connection can be used for executing queries. i.e. can connect to trino, can pass authn and authz and execute queries. So if queuing adds latency it's not a problem. Same happens in every single database - e.g. Postgres with connection limits will fail the `isValid` check even though system is up.

A fake endpoint with different query execution path doesn't seem to address the use-case.

Reply:
> `isValid` is not something that gets called so frequently as to cause issues with queing.

A user may have a limit of concurrent queries (eg 1) and they may be executing a long running query. Other connections should still be valid.

> Same happens in every single database - e.g. Postgres with connection limits will fail the isValid check even though system is up.

connection limits  => no connection => failure, understood

> A fake endpoint with different query execution path doesn't seem to address the use-case.

there can be many reasons why a query cannot be run: resource groups, permissions, catalog deployed, workers not up, etc.
`SELECT 1` may or may not validate all of them. In fact, server could optimize these queries away within `StatementResource` and not go through planning and execution. (cool optimization, ain't it?)

Those attributes are not connection state though; isValid should return true when we have a working connection to the server.

In particular, sole check of isValid should not initiate the cluster scale up (should one have a cluster with scale to zero capability).

Reply:
Ok, that makes sense. Your arguments are convincing.